tensorflow_addons/image/resampler_ops_test.py

```diff
@@ -0,0 +1,254 @@
+        samples assumed to be defined at the corresponding pixel coordinates.
+      x: numpy array of shape `[warp_height, warp_width]` containing
+        x coordinates at which interpolation will be performed.
+      y: numpy array of shape `[warp_height, warp_width]` containing
+        y coordinates at which interpolation will be performed.
+    Returns:
+      Numpy array of shape `[warp_height, warp_width]` containing interpolated
+        values.
+    """
+    shape = x.shape
+    x = np.asarray(x) + 1
+    y = np.asarray(y) + 1
+    data = np.lib.pad(data, 1, "constant", constant_values=0)
+
+    x_0 = np.floor(x).astype(int)
+    x_1 = x_0 + 1
+    y_0 = np.floor(y).astype(int)
+    y_1 = y_0 + 1
+
+    x_0 = np.clip(x_0, 0, data.shape[1] - 1)
+    x_1 = np.clip(x_1, 0, data.shape[1] - 1)
+    y_0 = np.clip(y_0, 0, data.shape[0] - 1)
+    y_1 = np.clip(y_1, 0, data.shape[0] - 1)
+
+    i_a = data[y_0, x_0]
+    i_b = data[y_1, x_0]
+    i_c = data[y_0, x_1]
+    i_d = data[y_1, x_1]
+
+    w_a = (x_1 - x) * (y_1 - y)
+    w_b = (x_1 - x) * (y - y_0)
+    w_c = (x - x_0) * (y_1 - y)
+    w_d = (x - x_0) * (y - y_0)
+
+    samples = (w_a * i_a + w_b * i_b + w_c * i_c + w_d * i_d)
+    samples.reshape(shape)
+
+    return samples
+
+
+def _make_warp(batch_size, warp_height, warp_width, dtype):
+    """Creates batch of warping coordinates."""
+    x, y = np.meshgrid(
+        np.linspace(0, warp_width - 1, warp_width),
+        np.linspace(0, warp_height - 1, warp_height))
+    warp = np.concatenate((x.reshape([warp_height, warp_width, 1]),
+                           y.reshape([warp_height, warp_width, 1])), 2)
+    warp = np.tile(
+        warp.reshape([1, warp_height, warp_width, 2]), [batch_size, 1, 1, 1])
+    warp += np.random.randn(*warp.shape)
+    return warp.astype(dtype)
+
+
+@test_utils.run_all_in_graph_and_eager_modes
+class ResamplerTest(tf.test.TestCase):
+    def test_op_forward_pass_gpu_float32(self):
+        self._test_op_forward_pass(True, tf.float32, 1e-4)
+
+    def test_op_forward_pass_gpu_float64(self):
+        self._test_op_forward_pass(True, tf.float64, 1e-5)
+
+    def test_op_forward_pass_cpu_float16(self):
+        self._test_op_forward_pass(False, tf.float16, 1e-2)
+
+    def test_op_forward_pass_cpu_float32(self):
+        self._test_op_forward_pass(False, tf.float32, 1e-4)
+
+    def test_op_forward_pass_cpu_float64(self):
+        self._test_op_forward_pass(False, tf.float64, 1e-5)
+
+    def test_op_backward_pass_gpu_float32(self):
+        self._test_op_backward_pass(True, tf.float32, 1e-3)
+
+    def test_op_backward_pass_cpu_float16(self):
+        self._test_op_backward_pass(False, tf.float16, 1e-3)
+
+    def test_op_backward_pass_cpu_float32(self):
+        self._test_op_backward_pass(False, tf.float32, 1e-4)
+
+    def test_op_backward_pass_cpu_float64(self):
+        self._test_op_backward_pass(False, tf.float64, 1e-6)
+
+    def _test_op_forward_pass(self, on_gpu, dtype, tol):
```

Comment:
I doubt there might be some mistakes for this function. The operations could only be executed when `on_gpu` is `True`. To validate if GPU is available, you might need to do something like this:

```python
use_gpu = on_gpu and tf.test.is_gpu_available()
with test_utils.device(use_gpu):
    # Do something
```

Reply:
Thanks for the gpu related comments, @WindQAQ. I am not currently able to test wtih gpu machine.